src/FilBeam.sol

```diff
@@ -0,0 +1,210 @@
+    {
+        if (newEpoch == 0) revert InvalidEpoch();
+
+        DataSetUsage storage usage = dataSetUsage[dataSetId];
+
+        if (newEpoch <= usage.maxReportedEpoch) revert InvalidEpoch();
+
+        usage.cdnBytesUsed += cdnBytesUsed;
+        usage.cacheMissBytesUsed += cacheMissBytesUsed;
+        usage.maxReportedEpoch = newEpoch;
+
+        emit UsageReported(dataSetId, newEpoch, cdnBytesUsed, cacheMissBytesUsed);
+    }
+
+    function settleCDNPaymentRail(uint256 dataSetId) external {
+        _settleCDNPaymentRail(dataSetId);
+    }
+
+    function settleCDNPaymentRailBatch(uint256[] calldata dataSetIds) external {
+        for (uint256 i = 0; i < dataSetIds.length; i++) {
+            _settleCDNPaymentRail(dataSetIds[i]);
+        }
+    }
+
+    function _settleCDNPaymentRail(uint256 dataSetId) internal {
+        DataSetUsage storage usage = dataSetUsage[dataSetId];
+
+        if (usage.maxReportedEpoch == 0) revert DataSetNotInitialized();
+        if (usage.maxReportedEpoch <= usage.lastCDNSettlementEpoch) revert NoUsageToSettle();
+
+        uint256 fromEpoch = usage.lastCDNSettlementEpoch + 1;
+        uint256 toEpoch = usage.maxReportedEpoch;
+        uint256 cdnAmount = usage.cdnBytesUsed * cdnRatePerByte;
+
+        if (cdnAmount > 0) {
+            fwss.settleCDNPaymentRails(dataSetId, cdnAmount, 0);
+        }
+
+        usage.lastCDNSettlementEpoch = toEpoch;
+        usage.cdnBytesUsed = 0;
+
+        emit CDNSettlement(dataSetId, fromEpoch, toEpoch, cdnAmount);
+    }
+
+    function settleCacheMissPaymentRail(uint256 dataSetId) external {
+        _settleCacheMissPaymentRail(dataSetId);
+    }
+
+    function settleCacheMissPaymentRailBatch(uint256[] calldata dataSetIds) external {
+        for (uint256 i = 0; i < dataSetIds.length; i++) {
+            _settleCacheMissPaymentRail(dataSetIds[i]);
+        }
+    }
+
+    function _settleCacheMissPaymentRail(uint256 dataSetId) internal {
+        DataSetUsage storage usage = dataSetUsage[dataSetId];
+
+        if (usage.maxReportedEpoch == 0) revert DataSetNotInitialized();
+        if (usage.maxReportedEpoch <= usage.lastCacheMissSettlementEpoch) revert NoUsageToSettle();
+
+        uint256 fromEpoch = usage.lastCacheMissSettlementEpoch + 1;
+        uint256 toEpoch = usage.maxReportedEpoch;
+        uint256 cacheMissAmount = usage.cacheMissBytesUsed * cacheMissRatePerByte;
+
+        if (cacheMissAmount > 0) {
+            fwss.settleCDNPaymentRails(dataSetId, 0, cacheMissAmount);
+        }
+
+        usage.lastCacheMissSettlementEpoch = toEpoch;
+        usage.cacheMissBytesUsed = 0;
+
+        emit CacheMissSettlement(dataSetId, fromEpoch, toEpoch, cacheMissAmount);
+    }
+
+    function terminateCDNPaymentRails(uint256 dataSetId) external onlyFilBeamController {
+        fwss.terminateCDNPaymentRails(dataSetId);
+
+        emit PaymentRailsTerminated(dataSetId);
+    }
+
+    function setFilBeamController(address _filBeamController) external onlyOwner {
```

Comment:
strange, I thought this contract was the controller contract? How can we change the controller's controller? Or is that intended, as there needs to be something controlling the controller?

Reply:
Ugh, yeah it's confusing. 

So this contract is an operator and it has it's own controller. This contract acts as a controller in case of FWSS. I think a good thing would be to:

- Rename this contract to `FilBeamOperator`
- Leave the controller address for `FilBeamOperator`
- Rename `filBeamController` (or whatever it's name is) inside FWSS to `filBeamOperator`
- Add method for migrating `filBeamOperator` on FWSS to this contract

Reply:
I suggest to sync on this terminology with FilOz, especially Jenni & Zen